### PR TITLE
feat/cody: Prevent Enterprise users from logging into dotcom/PLG (CODY-4295)

### PR DESCRIPTION
**Pull request by julialeex**
This PR does [Create mechanism to kick PLG logins/signup for user with email domains that below to enterprise customers](https://linear.app/sourcegraph/issue/CODY-4295/create-mechanism-to-kick-plg-loginssignup-for-user-with-email-domains).

**Problem:** We want to prevent enterprise customers from logging into PLG/dotcom.

This PR checks whether the user is an enterprise customer using `UserShouldUseEnterprise`, and whether the user tries to login into PLG/dotom through `isDotCom`, and logs the enterprise user out of PLG logins.

**Things this PR does:**

- `lib/shared/src`: Add a new field `userShouldUseEnterprise: boolean` to CodyClientConfig. Add `fetchConfigWithToken` to get this value before the user is logged in. A small refactor on `UnauthenticatedAuthStatus`.
- `vscode/src/auth`: If an enterprise user tries to sign in, auth validation flow will detect this and the sign in process would fail with an error message.
- `vscode/src/jsonrpc` and `vscode/src/chat`: Refactor on `UnauthenticatedAuthStatus`.
- `vscode/src/services`: Observer for ClientConfig changes.
- `vscode/webviews`: Sign out and show an error message if the user should use enterprise and is already logged into PLG.
- `vscode/test`: Tests.

## Test plan
**E2E test** 
Added test `test enterprise customers cannot log into dotcomUrl` in `auth.test.ts`.

**Test with debugger**
1. An enterprise user gets logged out for PLG/dotcom logins.
2. An enterprise user that is already logged into PLG/dotcom gets logged out.

https://github.com/user-attachments/assets/5a8e0bcc-d2f9-4cc0-a33b-9556f15203ad

https://github.com/user-attachments/assets/b354020b-0e3a-4627-912c-04c4373c64f1

https://github.com/user-attachments/assets/4428689c-1009-4e07-bcff-47bf02333edc

## Changelog
- Prevent PLG login methods for enterprise users

### CODE REVIEW

File: vscode/src/chat/chat-view/ChatController.ts [hunk 1]
```diff
@@ -553,6 +553,9 @@ export class ChatController implements vscode.Disposable, vscode.WebviewViewProv
         let clientConfig: CodyClientConfig | undefined
         try {
             clientConfig = await ClientConfigSingleton.getInstance().getConfig(abortController.signal)
+            if (clientConfig?.userShouldUseEnterprise && isDotCom(authStatus.endpoint)) {
+                signOut(authStatus.endpoint)
+            }
```
Comment (jamesmcnamara) [inline vscode/src/chat/chat-view/ChatController.ts:558]: I take your point on this now that I've seen it in action, and I think I gave you bad advice. The flash of the logged in screen before we're logged out looks janky and this code path is hard to follow for the next developer.

I think you're right that we should directly refresh this client config as part of the auth validation flow and stop the login with an error notification. I expect you'll need to add some methods to ClientConfigSingleton to achieve this.

File: vscode/test/e2e/auth.test.ts [hunk 2]
```diff
@@ -1,24 +1,24 @@
...
-        'cody.auth.logout:clicked',
+        'cody.signInNotification:shown',
+        'cody.auth:failed',
+        'cody.auth.login:firstEver',
         'cody.auth:disconnected',
         'cody.interactiveTutorial:attemptingStart',
         'cody.experiment.interactiveTutorial:enrolled',
     ],
-})('requires a valid auth token and allows logouts', async ({ page, sidebar }) => {
+})('requires a valid auth token and allows logouts', async ({ page, sidebar, nap }) => {
```
Comment (jamesmcnamara) [inline vscode/test/e2e/auth.test.ts:21]:
```suggestion
})('requires a valid auth token and allows logouts', async ({ page, sidebar }) => {
```

**Review by jamesmcnamara — CHANGES_REQUESTED** [review]
Yeah I definitely gave you some bad advice here :) I left some comments in line but basically, we should: 
- [x] Add a direct call to check the client config singleton from the auth flow. 
- [x] Show an error notification when the flag is false
- [x] Fix the tests so that we only set userShouldUseEnterprise to true in tests where we want it

File: vscode/src/chat/chat-view/ChatController.ts (hunk 1 above)
Comment (julialeex) [reply vscode/src/chat/chat-view/ChatController.ts:558]: 1. To summarize, we want to add the logic to the auth validation flow, before we log in (like how we were checking userInfo in in the auth validation flow before)? 
> The flash of the logged in screen before we're logged out looks janky
2. I thought we have to log in first because of this comment on the [ticket](https://linear.app/sourcegraph/issue/CODY-4295/create-mechanism-to-kick-plg-loginssignup-for-user-with-email-domains#comment-2aee6292)

Comment (jamesmcnamara) [reply vscode/src/chat/chat-view/ChatController.ts:558]: 1. Yes exactly.
2. I think that's more trying to capture that if a user is already logged in, and we refresh the client config and they are now supposed to be on an enterprise account, we should log them out. Which is a good point that that is another TODO.

Comment (julialeex) [reply vscode/src/chat/chat-view/ChatController.ts:558]: Got it, thanks for the clarification! Will update the code

Comment (julialeex) [reply vscode/src/chat/chat-view/ChatController.ts:558]: 1. I read the code for a bit more yesterday, and found that clientConfig is not populated until the user is authenticated (see [code1](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/cody/-/blob/lib/shared/src/sourcegraph-api/clientConfig.ts?L77) and [code2](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/cody/-/blob/lib/shared/src/sourcegraph-api/graphql/client.ts?L1575-1578))
2. Because of what I found in 1), I double checked the feature requirements in the [ticket](https://linear.app/sourcegraph/issue/CODY-4295/create-mechanism-to-kick-plg-loginssignup-for-user-with-email-domains#comment-868d5cc5) and it turns out we do want to log in first and then log the user out. If this is true, I have a TODO of adding the error message after we log the user out

File: vscode/test/e2e/auth.test.ts [hunk 3]
```diff
@@ -53,3 +57,57 @@ test.extend<ExpectedV2Events>({
...
+    })('test enterprise customers cannot log into dotcomUrl', async ({ page, sidebar }) => {
+    await expect(sidebar!.getByText('Sign in to Sourcegraph')).toBeVisible()
+    await sidebar!.getByRole('button', { name: 'Sourcegraph logo Continue' }).click()
+    await sidebar!.getByText('Sourcegraph Instance URL').click()
+    await sidebar!.getByPlaceholder('Example: https://instance.').click()
+    await sidebar!.getByPlaceholder('Example: https://instance.').fill(SERVER_URL)
+
+    await sidebar!.getByText('Access Token (Optional)').click()
+    await sidebar!.getByPlaceholder('Access token...').fill(VALID_TOKEN)
+    await sidebar!.getByRole('button', { name: 'Sign In' }).click()
```
Comment (jamesmcnamara) [inline vscode/test/e2e/auth.test.ts:87]: Great test! The risk with this is that the signin code updates a lot (or at least is in the process of a big migration), so any change to that flow will require this test to be updated as well. I think we can just update the sidebar signin like 
```ts
interface SidebarSigninOptions {
    disableNotifications: boolean
    skipAssertions: boolean
}

// Sign into Cody with valid auth from the sidebar
export const sidebarSignin = async (
    page: Page,
    sidebar: Frame | null,
    options: Partial<SidebarSigninOptions> = {}
): Promise<void> => {
...
}
```
And then you can just use the standard signin flow without asserting that it works.

File: vscode/test/e2e/auth.test.ts [hunk 4]
```diff
@@ -53,3 +57,57 @@ test.extend<ExpectedV2Events>({
...
+    await focusSidebar(page)
+
+    // Makes sure the sign in page is loaded in the sidebar view with Cody: Chat as the heading
+    // instead of the chat panel.
+    const sidebarFrame = getChatSidebarPanel(page)
+    await expect(sidebarFrame.getByText('Sign in to Sourcegraph')).toBeVisible()
+    await expect(page.getByRole('heading', { name: 'Cody: Chat' })).toBeVisible()
+    // Expect status bar to show the sign in button.
+    await expect(page.getByRole('button', { name: 'cody-logo-heavy Sign In, Sign' })).toBeVisible()
+})
```
Comment (jamesmcnamara) [inline vscode/test/e2e/auth.test.ts:113]: It looks like we're still missing a test that checks when the user is signed in and then we update the userShouldUseEnterprise to true and we test that they're logged out.

**Review by jamesmcnamara — CHANGES_REQUESTED** [review]
Looking good! The blockers are just not running the auth check unless it's dotcom and a test for logging out if the flag updates. Let me know if I can help with that!

File: vscode/src/auth/auth.ts [hunk 5]
```diff
@@ -451,6 +465,20 @@ export async function validateCredentials(
...
+
+    if (clientConfig?.userShouldUseEnterprise && isDotCom(config.auth.serverEndpoint)) {
+        return {
+            authenticated: false,
+            endpoint: config.auth.serverEndpoint,
+            pendingValidation: false,
+            userEnterprise: getEnterpriseName(userInfo.primaryEmail?.email || ''),
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (julialeex) [reply vscode/src/auth/auth.ts:481]: This is a great point, nice catch!

File: lib/shared/src/auth/types.ts [hunk 6]
```diff
@@ -48,6 +48,7 @@ export interface UnauthenticatedAuthStatus {
     showNetworkError?: boolean
     showInvalidAccessTokenError?: boolean
     pendingValidation: boolean
+    userEnterprise?: string
```
[the remark this answers is not in the document]
Comment (julialeex) [reply lib/shared/src/auth/types.ts:51]: This is a great suggestion and I did miss some places that are checking showNetworkError and showInvalidAccessTokenError! Since this is a refactor on the `UnauthenticatedAuthStatus`, would it be better to move the refactor to another PR, merge the refactor PR before this PR? This way, the current PR can focus just on the feature implementation.

[the remark this answers is not in the document]
Comment (jamesmcnamara) [reply lib/shared/src/auth/types.ts:51]: I don't think the refactor will be big enough to really overshadow this PR. It would have been more complicated if JetBrains or Eclipse used this value, but a quick grep seems to show that it doesn't.

I'd do it in this PR and if we think it's too big, we can split it out then.

File: vscode/test/e2e/auth.test.ts (hunk 4 above)
Comment (julialeex) [reply vscode/test/e2e/auth.test.ts:81]: Added the test!

File: vscode/src/auth/auth.ts [hunk 7]
```diff
@@ -282,17 +285,27 @@ async function showAuthResultMessage(
...
+        await vscode.window.showErrorMessage(
+            `Based on your email address we think you may be an employee of ${authStatus?.error?.enterprise}.
+            To get access to all your features please sign in through your organization\'s enterprise instance instead.
+            If you need assistance please contact your Sourcegraph admin.`
+        )
+    } else {
+        await vscode.window.showErrorMessage(
+            `Authentication failed. Please ensure Cody is enabled for ${authority} and verify your email address if required.`
+        )
+    }
```
Comment (jamesmcnamara) [inline vscode/src/auth/auth.ts:307]:
```suggestion
    if (authStatus?.error) {
        await vscode.window.showErrorMessage(getAuthErrorMessage(authStatus.error).message)
    }
```

File: vscode/src/services/AuthProvider.ts [hunk 8]
```diff
@@ -65,6 +94,36 @@ class AuthProvider implements vscode.Disposable {
...
+                            auth: {
+                                serverEndpoint: resolvedConfig.auth.serverEndpoint,
+                                accessToken: resolvedConfig.auth.accessToken,
+                            },
+                            clientState: {
+                                anonymousUserID: resolvedConfig.clientState.anonymousUserID,
+                            },
+                        }
+                        await this.validateAndUpdateAuthStatus(credentials, signal)
+                    })
```
Comment (jamesmcnamara) [inline vscode/src/services/AuthProvider.ts:122]:
```suggestion
                    abortableOperation(async (_, signal) => 
                        this.validateAndUpdateAuthStatus(await currentResolvedConfig(), signal)
                    )
```

File: vscode/webviews/App.tsx [hunk 9]
```diff
@@ -167,9 +167,11 @@ export const App: React.FunctionComponent<{ vscodeAPI: VSCodeWrapper }> = ({ vsc
         <ComposedWrappers wrappers={wrappers}>
             {view === View.Login || !config.authStatus.authenticated ? (
                 <div className={styles.outerContainer}>
-                    {!config.authStatus.authenticated && config.authStatus.showNetworkError && (
-                        <ConnectivityStatusBanner />
-                    )}
+                    {!config.authStatus.authenticated &&
+                        (config.authStatus.error?.type === 'network-error' ||
+                            config.authStatus.error?.type === 'enterprise-user-logged-into-dotcom') && (
```
Comment (jamesmcnamara) [inline vscode/webviews/App.tsx:172]:
```suggestion
                        (config.authStatus.error) && (
```

File: vscode/webviews/components/AuthenticationErrorBanner.tsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+                    <span className="tw-font-bold">Network Error</span> &mdash; Cody is unreachable
+                </span>
+            )}
+            {error.type === 'enterprise-user-logged-into-dotcom' && (
+                <span className="tw-flex-1 tw-text-xs">
+                    Based on your email address we think you may be an employee of {error.enterprise}. To
+                    get access to all your features please sign in through your organization's enterprise
+                    instance instead. If you need assistance please contact your Sourcegraph admin.
+                </span>
+            )}
```
Comment (jamesmcnamara) [inline vscode/webviews/components/AuthenticationErrorBanner.tsx:31]:
```suggestion
            <span className="tw-flex-1 tw-text-xs">
                {errorMsg.title && (
                    <>
                        <span className="tw-font-bold">{errorMsg.title}</span> &mdash;
                    </>
                )}
                {errorMsg.message}
            </span>
```

And then just add 
```ts
const errorMsg = getAuthErrorMessage(error)
```
to the component.

**Review by jamesmcnamara — APPROVED** [review]
Still some cleanup but this works! Great stuff!

**Review by dominiccooney — APPROVED** [review]
This looks good to me.

We should create some testing accounts that will trigger this, and have QA test this continually in JetBrains and VSCode (...at least until we can add an automated JetBrains test.)

File: agent/src/auth.test.ts [hunk 11]
```diff
@@ -50,6 +50,7 @@ describe(
         beforeAll(async () => {
             await workspace.beforeAll()
             await client.beforeAll()
+            await new Promise(resolve => setTimeout(resolve, 500))
```
Comment (pkukielka) [inline agent/src/auth.test.ts:53]: That looks suspicious. Why is that needed?

**Review by dominiccooney — APPROVED** [review]
Still LGTM.

Now that the authStatus/clientConfig are decoupled like that, do we need the timeouts in the tests?
